pylint/checkers/variables.py

```diff
@@ -1230,9 +1230,7 @@ def visit_functiondef(self, node: nodes.FunctionDef) -> None:
                 # Do not take in account redefined names for the purpose
                 # of type checking.:
                 if any(
-                    isinstance(definition.parent, nodes.If)
-                    and definition.parent.test.as_string() in TYPING_TYPE_CHECKS_GUARDS
-                    for definition in globs[name]
+                    in_type_checking_block(definition) for definition in globs[name]
```

Comment:
Not sure if this is in scope, but there is only one reference to `TYPING_TYPE_CHECKS_GUARDS` remaining. Should we remove it as well and use this function everywhere?

Reply:
I think that should be a follow-up PR. At first glance I'm not sure there's a drop-in replacement.

Reply:
```diff
diff --git a/pylint/checkers/variables.py b/pylint/checkers/variables.py
index 65a9e8c8b..59e9f7fba 100644
--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -2011,10 +2011,7 @@ class VariablesChecker(BaseChecker):
             if isinstance(defstmt, (nodes.Import, nodes.ImportFrom)):
                 defstmt_parent = defstmt.parent
 
-                if (
-                    isinstance(defstmt_parent, nodes.If)
-                    and defstmt_parent.test.as_string() in TYPING_TYPE_CHECKS_GUARDS
-                ):
+                if in_type_checking_block(defstmt):
                     # Exempt those definitions that are used inside the type checking
                     # guard or that are defined in both type checking guard branches.
                     used_in_branch = defstmt_parent.parent_of(node)

```

Passed all functional tests for me locally.

Reply:
I'm concerned there aren't enough tests. This example depends on `defstmt_parent` being `nodes.If`. `in_type_checking_block` is more robust, and doesn't assume direct parentage. There's more control flow in this example; it needs more thought.

Reply:
Ah yeah, my bad. What about:
```
if isinstance(defstmt_parent, nodes.If) and in_type_checking_block(defstmt):
```

The `in_type_checking_block` doesn't do any control flow analysis and only makes sure we infer instead of relying on the string representation. As far as I see it is a drop-in replacement for `defstmt_parent.test.as_string() in TYPING_TYPE_CHECKS_GUARDS`. (But not for `isinstance(defstmt_parent, nodes.If)` as you correctly spotted!)

Reply:
I still don't think it's a drop-in replacement because `in_type_checking_block()` is recursive. We don't know which `if` ancestor met the condition, but the way this block is written, the assumption is that `defstmt_parent` is the one that met the condition, and doing a drop-in replacement will violate that condition. Leading me to think it needs more refactoring and a follow-up issue.

Reply:
Fair enough. Shall we open an issue/todo for this? I'm also fine with leaving as is.

Reply:
No problem, I'll do it. 👍🏻

Reply:
#7539